src/listing.rs

```diff
@@ -235,46 +235,49 @@ pub fn directory_listing<S>(
 
     let color_scheme = query_params.theme.unwrap_or(default_color_scheme);
 
-    if let Some(compression_method) = &query_params.download {
+    if let Some(compression_method) = query_params.download {
         log::info!(
             "Creating an archive ({extension}) of {path}...",
             extension = compression_method.extension(),
             path = &dir.path.display().to_string()
         );
-        match compression_method.create_archive(&dir.path, skip_symlinks) {
-            Ok((filename, content)) => {
-                log::info!("{file} successfully created !", file = &filename);
-                Ok(HttpResponse::Ok()
-                    .content_type(compression_method.content_type())
-                    .content_encoding(compression_method.content_encoding())
-                    .header("Content-Transfer-Encoding", "binary")
-                    .header(
-                        "Content-Disposition",
-                        format!("attachment; filename={:?}", filename),
-                    )
-                    .chunked()
-                    .body(Body::Streaming(Box::new(once(Ok(content))))))
-            }
-            Err(err) => {
-                errors::log_error_chain(err.to_string());
-                Ok(HttpResponse::Ok()
-                    .status(http::StatusCode::INTERNAL_SERVER_ERROR)
-                    .body(
-                        renderer::render_error(
-                            &err.to_string(),
-                            StatusCode::INTERNAL_SERVER_ERROR,
-                            serve_path,
-                            query_params.sort,
-                            query_params.order,
-                            color_scheme,
-                            default_color_scheme,
-                            false,
-                            true,
-                        )
-                        .into_string(),
-                    ))
+
+        let filename = format!(
+            "{}.{}",
+            dir.path.file_name().unwrap().to_str().unwrap(),
+            compression_method.extension()
+        );
+
+        // Create a pipe to connect the archive creation thread and the response.
+        // Include 10 messages of buffer for erratic connection speeds.
+        let (tx, rx) = futures::sync::mpsc::channel(10);
+        let pipe = crate::pipe::Pipe::new(tx);
+
+        // Start the actual archive creation in a separate thread.
+        let dir = dir.path.to_path_buf();
+        std::thread::spawn(move || {
+            if let Err(err) = compression_method.create_archive(dir, skip_symlinks, pipe) {
+                log::error!("Error during archive creation: {:?}", err);
             }
-        }
+        });
+
+        // `<rx as Stream>::Error == ()` but we want `actix_web::error::Error`
+        let rx = rx.map_err(|_| {
+            ContextualError::CustomError("pipes never fail".to_string())
+                .compat()
+                .into()
```

Comment:
Note that I'm not super familiar with error conversions, so there may be an easier way to create an actix error.

Reply:
Using `unreachable!` instead. [`mpsc::Receiver` never return errors](https://docs.rs/futures/0.1.27/src/futures/sync/mpsc/mod.rs.html#923-958).